text/0000-futures.md

```diff
@@ -0,0 +1,813 @@
+- An executor, which is the "default" place to spawn further tasks.
+
+Notably, this list does *not* include task-local data; that can be addressed
+externally, as we'll see in a later section.
+
+The `core::task::Context` type gathers the (stack-rooted) contextual information
+together, and is passed by mutable reference to all polling functions:
+
+```rust
+/// Information about the currently-running task.
+///
+/// Contexts are always tied to the stack, since they are set up specifically
+/// when performing a single `poll` step on a task.
+pub struct Context<'a> { .. }
+
+impl<'a> Context<'a> {
+    /// Note: this signature is future-proofed for `E: Executor` later.
+    pub fn new<E>(waker: &'a Waker, executor: &'a mut E) -> Context<'a>
+        where E: BoxExecutor;
+
+    /// Get the `Waker` associated with the current task.
+    pub fn waker(&self) -> &Waker
+
+    /// Spawn a future onto the default executor.
+    ///
+    /// # Panics
+    ///
+    /// This method will panic if the default executor is unable to spawn.
+    /// To handle executor errors, use the `executor` method instead.
+    pub fn spawn(&mut self, f: impl Future<Item = ()> + 'static + Send);
+
+    /// Get the default executor associated with this task.
+    ///
+    /// This method is useful primarily if you want to explicitly handle
+    /// spawn failures.
+    ///
+    /// NB: this will remain unstable until the final `Executor` trait is ready.
+    pub fn executor(&mut self) -> &mut BoxExecutor;
+}
+```
+
+Note that the `spawn` method here will box until `Executor` is added.
+
+## `core::future` module
+
+With all of the above task infrastructure in place, defining `Future` is
+straightforward.  Like `Iterator`, the `Future` trait has a single required
+method, `poll`, and a large number of provided methods (called "adapters" or
+"combinators"). We'll look first at `poll`:
+
+```rust
+pub trait Future {
+    /// The type of value produced on completion.
+    type Item;
+
+    /// Attempt to resolve the future to a final value, registering
+    /// the current task for wakeup if the value is not yet available.
+    ///
+    /// # Return value
+    ///
+    /// This function returns:
+    ///
+    /// - `Poll::Pending` if the future is not ready yet.
+    /// - `Poll::Ready(val)` with the result `val` of this future if it completed.
+    ///
+    /// Once a future has completed, clients should not `poll` it again.
+    ///
+    /// When a future is not ready yet, `poll` returns `Poll::Pending`.
+    /// The future will *also* register the interest of the current task in the
+    /// value being produced. For example, if the future represents the availability
+    /// of data on a socket, then the task is recorded so that when data arrives,
+    /// it is woken up (via `cx.waker()`). Once a task has been woken up,
+    /// it should attempt to `poll` the future again, which may or may not
+    /// produce a final value at that time.
```

Comment:
I'd love to have it made explicit that it's not an error to poll a future even before a registered 'interest' has caused a task wakeup.

Reply:
What would be the use case for polling a future multiple times like that?

(Intuitively, I would expect that assuming that it doesn't happen allows for some implementation simplification, e.g. one doesn't need to guard against multiple wakeups being scheduled for a single I/O event.)

Reply:
The main one I can think of is implementing `select` and `join`, you don’t know which of your sub-futures caused the wakeup so you have to poll them all to see if any are ready yet.

Reply:
I would think that such O(N) "poll all the inner futures" behaviour would be considered problematic. Hasn't there been some work to help a future figure out the cause of a wakeup?

Reply:
IMO `select` and `join` should only be used for a very small number of futures, e.g. selecting a single future with a timeout to cancel it. If you have many sub-parts then these should be `spawn`ed off to separate tasks (and so be woken separately by the `Executor`) and a more efficient way to notify when all/one is complete should be used.